File: code_julia/equilibria/interior.jl
#=============================================================================
Author: Rayla Kurosaki
GitHub: https://github.com/rkp1503
=============================================================================#

module interior

import Polynomials: Polynomial, roots

function get_equilibria(parameter_values)
    equilibria = get_analytical_solutions(parameter_values)
    r₂₁, r₃₁, p, ϕ₁₂, ϕ₂₁, ϕ₁₃, u₁, u₂, u₃, u₄ = parameter_values
    for (i, equilibrium) in enumerate(equilibria)
        x, y, z = equilibrium
        # if !((x > 0) && (z > 0))
        #     deleteat!(equilibria, i)
        # end
        cond1 = (1+ϕ₁₂*y^2)/ϕ₁₃ > z
        cond2 = y > (u₂*(u₄-r₃₁))/((u₃-u₄+r₃₁)*(1-p))
        if !all([cond1, cond2])
            deleteat!(equilibria, i)
        end
    end
    return equilibria
end;

function get_stable_equilibria!(equilibria, parameter_values)
    r₂₁, r₃₁, p, ϕ₁₂, ϕ₂₁, ϕ₁₃, u₁, u₂, u₃, u₄ = parameter_values
    for (i, equilibrium) in enumerate(equilibria)
        x, y, z = equilibrium
        j₁₁ = 1-2*x+ϕ₁₂*y^2-ϕ₁₃*z
        j₁₂ = 2*ϕ₁₂*x*y
        j₁₃ = -ϕ₁₃*x
        j₂₁ = 2*ϕ₂₁*r₂₁*x*y
        j₂₂ = r₂₁*(1-2*y+ϕ₂₁*x^2)-((u₁*u₂*(1-p)*z)/(u₂+(1-p)*y)^2)
        j₂₃ = -((u₁*(1-p)*y)/(u₂+(1-p)*y))
        j₃₂ = (u₂*u₃*(1-p)*z)/((u₂+(1-p)*y)^2)
        j₃₃ = r₃₁*(1-2*z)-u₄+((u₃*(1-p)*y)/(u₂+(1-p)*y))
        C₂ = -j₁₁-j₂₂-j₃₃
        C₁ = j₁₁*j₂₂+j₁₁*j₃₃+j₂₂*j₃₃-j₁₂*j₂₁-j₂₃*j₃₂
        C₀ = (j₁₁*j₂₃*j₃₂+j₁₂*j₂₁*j₃₃)-(j₁₁*j₂₂*j₃₃+j₁₃*j₂₁*j₃₂)
        cond1 = C₂ > 0
        cond2 = C₁ > 0
        cond3 = C₀ > 0
        cond4 = C₂*C₁ > C₀
        if !all([cond1, cond2, cond3, cond4])
            deleteat!(equilibria, i)
        end
    end
    return nothing
end;

function get_analytical_solutions(parameter_values)
    sols_real = []
    r₂₁, r₃₁, p, ϕ₁₂, ϕ₂₁, ϕ₁₃, u₁, u₂, u₃, u₄ = parameter_values
    Y₆ = r₂₁*r₃₁^2*ϕ₁₂^2*ϕ₂₁*(1-p)^2
    Y₅ = 2*r₂₁*r₃₁^2*u₂*ϕ₁₂^2*ϕ₂₁*(1-p)
    Y₄ = -2*p^2*r₂₁*r₃₁^2*ϕ₁₂*ϕ₁₃*ϕ₂₁+2*p^2*r₂₁*r₃₁^2*ϕ₁₂*ϕ₂₁-2*p^2*r₂₁*r₃₁*u₃*ϕ₁₂*ϕ₁₃*ϕ₂₁+2*p^2*r₂₁*r₃₁*u₄*ϕ₁₂*ϕ₁₃*ϕ₂₁+4*p*r₂₁*r₃₁^2*ϕ₁₂*ϕ₁₃*ϕ₂₁-4*p*r₂₁*r₃₁^2*ϕ₁₂*ϕ₂₁+4*p*r₂₁*r₃₁*u₃*ϕ₁₂*ϕ₁₃*ϕ₂₁-4*p*r₂₁*r₃₁*u₄*ϕ₁₂*ϕ₁₃*ϕ₂₁+r₂₁*r₃₁^2*u₂^2*ϕ₁₂^2*ϕ₂₁-2*r₂₁*r₃₁^2*ϕ₁₂*ϕ₁₃*ϕ₂₁+2*r₂₁*r₃₁^2*ϕ₁₂*ϕ₂₁-2*r₂₁*r₃₁*u₃*ϕ₁₂*ϕ₁₃*ϕ₂₁+2*r₂₁*r₃₁*u₄*ϕ₁₂*ϕ₁₃*ϕ₂₁
    Y₃ = -p^2*r₂₁*r₃₁^2+4*p*r₂₁*r₃₁^2*u₂*ϕ₁₂*ϕ₁₃*ϕ₂₁-4*p*r₂₁*r₃₁^2*u₂*ϕ₁₂*ϕ₂₁+2*p*r₂₁*r₃₁^2+2*p*r₂₁*r₃₁*u₂*u₃*ϕ₁₂*ϕ₁₃*ϕ₂₁-4*p*r₂₁*r₃₁*u₂*u₄*ϕ₁₂*ϕ₁₃*ϕ₂₁-4*r₂₁*r₃₁^2*u₂*ϕ₁₂*ϕ₁₃*ϕ₂₁+4*r₂₁*r₃₁^2*u₂*ϕ₁₂*ϕ₂₁-r₂₁*r₃₁^2-2*r₂₁*r₃₁*u₂*u₃*ϕ₁₂*ϕ₁₃*ϕ₂₁+4*r₂₁*r₃₁*u₂*u₄*ϕ₁₂*ϕ₁₃*ϕ₂₁
    Y₂ = p^2*r₂₁*r₃₁^2*ϕ₁₃^2*ϕ₂₁-2*p^2*r₂₁*r₃₁^2*ϕ₁₃*ϕ₂₁+p^2*r₂₁*r₃₁^2*ϕ₂₁+p^2*r₂₁*r₃₁^2+2*p^2*r₂₁*r₃₁*u₃*ϕ₁₃^2*ϕ₂₁-2*p^2*r₂₁*r₃₁*u₃*ϕ₁₃*ϕ₂₁-2*p^2*r₂₁*r₃₁*u₄*ϕ₁₃^2*ϕ₂₁+2*p^2*r₂₁*r₃₁*u₄*ϕ₁₃*ϕ₂₁+p^2*r₂₁*u₃^2*ϕ₁₃^2*ϕ₂₁-2*p^2*r₂₁*u₃*u₄*ϕ₁₃^2*ϕ₂₁+p^2*r₂₁*u₄^2*ϕ₁₃^2*ϕ₂₁+2*p*r₂₁*r₃₁^2*u₂-2*p*r₂₁*r₃₁^2*ϕ₁₃^2*ϕ₂₁+4*p*r₂₁*r₃₁^2*ϕ₁₃*ϕ₂₁-2*p*r₂₁*r₃₁^2*ϕ₂₁-2*p*r₂₁*r₃₁^2-4*p*r₂₁*r₃₁*u₃*ϕ₁₃^2*ϕ₂₁+4*p*r₂₁*r₃₁*u₃*ϕ₁₃*ϕ₂₁+4*p*r₂₁*r₃₁*u₄*ϕ₁₃^2*ϕ₂₁-4*p*r₂₁*r₃₁*u₄*ϕ₁₃*ϕ₂₁-2*p*r₂₁*u₃^2*ϕ₁₃^2*ϕ₂₁+4*p*r₂₁*u₃*u₄*ϕ₁₃^2*ϕ₂₁-2*p*r₂₁*u₄^2*ϕ₁₃^2*ϕ₂₁-2*r₂₁*r₃₁^2*u₂^2*ϕ₁₂*ϕ₁₃*ϕ₂₁+2*r₂₁*r₃₁^2*u₂^2*ϕ₁₂*ϕ₂₁-2*r₂₁*r₃₁^2*u₂+r₂₁*r₃₁^2*ϕ₁₃^2*ϕ₂₁-2*r₂₁*r₃₁^2*ϕ₁₃*ϕ₂₁+r₂₁*r₃₁^2*ϕ₂₁+r₂₁*r₃₁^2+2*r₂₁*r₃₁*u₂^2*u₄*ϕ₁₂*ϕ₁₃*ϕ₂₁+2*r₂₁*r₃₁*u₃*ϕ₁₃^2*ϕ₂₁-2*r₂₁*r₃₁*u₃*ϕ₁₃*ϕ₂₁-2*r₂₁*r₃₁*u₄*ϕ₁₃^2*ϕ₂₁+2*r₂₁*r₃₁*u₄*ϕ₁₃*ϕ₂₁+r₂₁*u₃^2*ϕ₁₃^2*ϕ₂₁-2*r₂₁*u₃*u₄*ϕ₁₃^2*ϕ₂₁+r₂₁*u₄^2*ϕ₁₃^2*ϕ₂₁
    Y₁ = -p^2*r₃₁^2*u₁-p^2*r₃₁*u₁*u₃+p^2*r₃₁*u₁*u₄-2*p*r₂₁*r₃₁^2*u₂*ϕ₁₃^2*ϕ₂₁+4*p*r₂₁*r₃₁^2*u₂*ϕ₁₃*ϕ₂₁-2*p*r₂₁*r₃₁^2*u₂*ϕ₂₁-2*p*r₂₁*r₃₁^2*u₂-2*p*r₂₁*r₃₁*u₂*u₃*ϕ₁₃^2*ϕ₂₁+2*p*r₂₁*r₃₁*u₂*u₃*ϕ₁₃*ϕ₂₁+4*p*r₂₁*r₃₁*u₂*u₄*ϕ₁₃^2*ϕ₂₁-4*p*r₂₁*r₃₁*u₂*u₄*ϕ₁₃*ϕ₂₁+2*p*r₂₁*u₂*u₃*u₄*ϕ₁₃^2*ϕ₂₁-2*p*r₂₁*u₂*u₄^2*ϕ₁₃^2*ϕ₂₁+2*p*r₃₁^2*u₁+2*p*r₃₁*u₁*u₃-2*p*r₃₁*u₁*u₄-r₂₁*r₃₁^2*u₂^2+2*r₂₁*r₃₁^2*u₂*ϕ₁₃^2*ϕ₂₁-4*r₂₁*r₃₁^2*u₂*ϕ₁₃*ϕ₂₁+2*r₂₁*r₃₁^2*u₂*ϕ₂₁+2*r₂₁*r₃₁^2*u₂+2*r₂₁*r₃₁*u₂*u₃*ϕ₁₃^2*ϕ₂₁-2*r₂₁*r₃₁*u₂*u₃*ϕ₁₃*ϕ₂₁-4*r₂₁*r₃₁*u₂*u₄*ϕ₁₃^2*ϕ₂₁+4*r₂₁*r₃₁*u₂*u₄*ϕ₁₃*ϕ₂₁-2*r₂₁*u₂*u₃*u₄*ϕ₁₃^2*ϕ₂₁+2*r₂₁*u₂*u₄^2*ϕ₁₃^2*ϕ₂₁-r₃₁^2*u₁-r₃₁*u₁*u₃+r₃₁*u₁*u₄
    Y₀ = p*r₃₁^2*u₁*u₂-p*r₃₁*u₁*u₂*u₄+r₂₁*r₃₁^2*u₂^2*ϕ₁₃^2*ϕ₂₁-2*r₂₁*r₃₁^2*u₂^2*ϕ₁₃*ϕ₂₁+r₂₁*r₃₁^2*u₂^2*ϕ₂₁+r₂₁*r₃₁^2*u₂^2-2*r₂₁*r₃₁*u₂^2*u₄*ϕ₁₃^2*ϕ₂₁+2*r₂₁*r₃₁*u₂^2*u₄*ϕ₁₃*ϕ₂₁+r₂₁*u₂^2*u₄^2*ϕ₁₃^2*ϕ₂₁-r₃₁^2*u₁*u₂+r₃₁*u₁*u₂*u₄
    if Y₀ < 0
        for root in roots(Polynomial([Y₀, Y₁, Y₂, Y₃, Y₄, Y₅, Y₆], :y))
            if (real(root) > 0) && (imag(root) == 0)
                y = real(root)
                z = 1+((((u₃*(1-p)*y)/(u₂+(1-p)*y))-u₄)/r₃₁)
                x = 1+ϕ₁₂*y^2-ϕ₁₃*z
                push!(sols_real, (x, y, z))
            end
        end
    end
    return sols_real
end;

function descarte(lst)
    sign_changes = 0
    for i in 1:(length(lst)-1)
        if lst[i]*lst[i+1] < 0
            sign_changes += 1
        end
    end
    return isodd(sign_changes)
end;

end

File: code_python/equilibria/interior.py
"""
Author: Rayla Kurosaki

GitHub: https://github.com/rkp1503
"""

import numpy as np


def get_equilibria(parameter_values):
    equilibria = get_analytical_solutions(parameter_values)
    r_yx, r_zx, p, phi_xy, phi_yx, phi_xz, u_1, u_2, u_3, u_4 = parameter_values
    for (i, equilibrium) in enumerate(equilibria):
        x, y, z = equilibrium
        # if not ((x > 0) and (z > 0)):
        #     del equilibria[i]
        #     pass
        cond1 = (1 + phi_xy * y ** 2) / phi_xz > z
        cond2 = y > (u_2 * (u_4 - r_zx)) / ((u_3 - u_4 + r_zx) * (1 - p))
        if not all([cond1, cond2]):
            del equilibria[i]
            pass
        pass
    return equilibria


def get_analytical_solutions(parameter_values):
    sols_real = []
    r_yx, r_zx, p, phi_xy, phi_yx, phi_xz, u_1, u_2, u_3, u_4 = parameter_values
    Y_6 = r_yx * r_zx ** 2 * phi_xy ** 2 * phi_yx * (1 - p) ** 2
    Y_5 = 2 * r_yx * r_zx ** 2 * u_2 * phi_xy ** 2 * phi_yx * (1 - p)
    Y_4 = -2 * p ** 2 * r_yx * r_zx ** 2 * phi_xy * phi_xz * phi_yx + 2 * p ** 2 * r_yx * r_zx ** 2 * phi_xy * phi_yx - 2 * p ** 2 * r_yx * r_zx * u_3 * phi_xy * phi_xz * phi_yx + 2 * p ** 2 * r_yx * r_zx * u_4 * phi_xy * phi_xz * phi_yx + 4 * p * r_yx * r_zx ** 2 * phi_xy * phi_xz * phi_yx - 4 * p * r_yx * r_zx ** 2 * phi_xy * phi_yx + 4 * p * r_yx * r_zx * u_3 * phi_xy * phi_xz * phi_yx - 4 * p * r_yx * r_zx * u_4 * phi_xy * phi_xz * phi_yx + r_yx * r_zx ** 2 * u_2 ** 2 * phi_xy ** 2 * phi_yx - 2 * r_yx * r_zx ** 2 * phi_xy * phi_xz * phi_yx + 2 * r_yx * r_zx ** 2 * phi_xy * phi_yx - 2 * r_yx * r_zx * u_3 * phi_xy * phi_xz * phi_yx + 2 * r_yx * r_zx * u_4 * phi_xy * phi_xz * phi_yx
    Y_3 = -p ** 2 * r_yx * r_zx ** 2 + 4 * p * r_yx * r_zx ** 2 * u_2 * phi_xy * phi_xz * phi_yx - 4 * p * r_yx * r_zx ** 2 * u_2 * phi_xy * phi_yx + 2 * p * r_yx * r_zx ** 2 + 2 * p * r_yx * r_zx * u_2 * u_3 * phi_xy * phi_xz * phi_yx - 4 * p * r_yx * r_zx * u_2 * u_4 * phi_xy * phi_xz * phi_yx - 4 * r_yx * r_zx ** 2 * u_2 * phi_xy * phi_xz * phi_yx + 4 * r_yx * r_zx ** 2 * u_2 * phi_xy * phi_yx - r_yx * r_zx ** 2 - 2 * r_yx * r_zx * u_2 * u_3 * phi_xy * phi_xz * phi_yx + 4 * r_yx * r_zx * u_2 * u_4 * phi_xy * phi_xz * phi_yx
    Y_2 = p ** 2 * r_yx * r_zx ** 2 * phi_xz ** 2 * phi_yx - 2 * p ** 2 * r_yx * r_zx ** 2 * phi_xz * phi_yx + p ** 2 * r_yx * r_zx ** 2 * phi_yx + p ** 2 * r_yx * r_zx ** 2 + 2 * p ** 2 * r_yx * r_zx * u_3 * phi_xz ** 2 * phi_yx - 2 * p ** 2 * r_yx * r_zx * u_3 * phi_xz * phi_yx - 2 * p ** 2 * r_yx * r_zx * u_4 * phi_xz ** 2 * phi_yx + 2 * p ** 2 * r_yx * r_zx * u_4 * phi_xz * phi_yx + p ** 2 * r_yx * u_3 ** 2 * phi_xz ** 2 * phi_yx - 2 * p ** 2 * r_yx * u_3 * u_4 * phi_xz ** 2 * phi_yx + p ** 2 * r_yx * u_4 ** 2 * phi_xz ** 2 * phi_yx + 2 * p * r_yx * r_zx ** 2 * u_2 - 2 * p * r_yx * r_zx ** 2 * phi_xz ** 2 * phi_yx + 4 * p * r_yx * r_zx ** 2 * phi_xz * phi_yx - 2 * p * r_yx * r_zx ** 2 * phi_yx - 2 * p * r_yx * r_zx ** 2 - 4 * p * r_yx * r_zx * u_3 * phi_xz ** 2 * phi_yx + 4 * p * r_yx * r_zx * u_3 * phi_xz * phi_yx + 4 * p * r_yx * r_zx * u_4 * phi_xz ** 2 * phi_yx - 4 * p * r_yx * r_zx * u_4 * phi_xz * phi_yx - 2 * p * r_yx * u_3 ** 2 * phi_xz ** 2 * phi_yx + 4 * p * r_yx * u_3 * u_4 * phi_xz ** 2 * phi_yx - 2 * p * r_yx * u_4 ** 2 * phi_xz ** 2 * phi_yx - 2 * r_yx * r_zx ** 2 * u_2 ** 2 * phi_xy * phi_xz * phi_yx + 2 * r_yx * r_zx ** 2 * u_2 ** 2 * phi_xy * phi_yx - 2 * r_yx * r_zx ** 2 * u_2 + r_yx * r_zx ** 2 * phi_xz ** 2 * phi_yx - 2 * r_yx * r_zx ** 2 * phi_xz * phi_yx + r_yx * r_zx ** 2 * phi_yx + r_yx * r_zx ** 2 + 2 * r_yx * r_zx * u_2 ** 2 * u_4 * phi_xy * phi_xz * phi_yx + 2 * r_yx * r_zx * u_3 * phi_xz ** 2 * phi_yx - 2 * r_yx * r_zx * u_3 * phi_xz * phi_yx - 2 * r_yx * r_zx * u_4 * phi_xz ** 2 * phi_yx + 2 * r_yx * r_zx * u_4 * phi_xz * phi_yx + r_yx * u_3 ** 2 * phi_xz ** 2 * phi_yx - 2 * r_yx * u_3 * u_4 * phi_xz ** 2 * phi_yx + r_yx * u_4 ** 2 * phi_xz ** 2 * phi_yx
    Y_1 = -p ** 2 * r_zx ** 2 * u_1 - p ** 2 * r_zx * u_1 * u_3 + p ** 2 * r_zx * u_1 * u_4 - 2 * p * r_yx * r_zx ** 2 * u_2 * phi_xz ** 2 * phi_yx + 4 * p * r_yx * r_zx ** 2 * u_2 * phi_xz * phi_yx - 2 * p * r_yx * r_zx ** 2 * u_2 * phi_yx - 2 * p * r_yx * r_zx ** 2 * u_2 - 2 * p * r_yx * r_zx * u_2 * u_3 * phi_xz ** 2 * phi_yx + 2 * p * r_yx * r_zx * u_2 * u_3 * phi_xz * phi_yx + 4 * p * r_yx * r_zx * u_2 * u_4 * phi_xz ** 2 * phi_yx - 4 * p * r_yx * r_zx * u_2 * u_4 * phi_xz * phi_yx + 2 * p * r_yx * u_2 * u_3 * u_4 * phi_xz ** 2 * phi_yx - 2 * p * r_yx * u_2 * u_4 ** 2 * phi_xz ** 2 * phi_yx + 2 * p * r_zx ** 2 * u_1 + 2 * p * r_zx * u_1 * u_3 - 2 * p * r_zx * u_1 * u_4 - r_yx * r_zx ** 2 * u_2 ** 2 + 2 * r_yx * r_zx ** 2 * u_2 * phi_xz ** 2 * phi_yx - 4 * r_yx * r_zx ** 2 * u_2 * phi_xz * phi_yx + 2 * r_yx * r_zx ** 2 * u_2 * phi_yx + 2 * r_yx * r_zx ** 2 * u_2 + 2 * r_yx * r_zx * u_2 * u_3 * phi_xz ** 2 * phi_yx - 2 * r_yx * r_zx * u_2 * u_3 * phi_xz * phi_yx - 4 * r_yx * r_zx * u_2 * u_4 * phi_xz ** 2 * phi_yx + 4 * r_yx * r_zx * u_2 * u_4 * phi_xz * phi_yx - 2 * r_yx * u_2 * u_3 * u_4 * phi_xz ** 2 * phi_yx + 2 * r_yx * u_2 * u_4 ** 2 * phi_xz ** 2 * phi_yx - r_zx ** 2 * u_1 - r_zx * u_1 * u_3 + r_zx * u_1 * u_4
    Y_0 = p * r_zx ** 2 * u_1 * u_2 - p * r_zx * u_1 * u_2 * u_4 + r_yx * r_zx ** 2 * u_2 ** 2 * phi_xz ** 2 * phi_yx - 2 * r_yx * r_zx ** 2 * u_2 ** 2 * phi_xz * phi_yx + r_yx * r_zx ** 2 * u_2 ** 2 * phi_yx + r_yx * r_zx ** 2 * u_2 ** 2 - 2 * r_yx * r_zx * u_2 ** 2 * u_4 * phi_xz ** 2 * phi_yx + 2 * r_yx * r_zx * u_2 ** 2 * u_4 * phi_xz * phi_yx + r_yx * u_2 ** 2 * u_4 ** 2 * phi_xz ** 2 * phi_yx - r_zx ** 2 * u_1 * u_2 + r_zx * u_1 * u_2 * u_4
    if Y_0 < 0:
        for root in np.roots([Y_6, Y_5, Y_4, Y_3, Y_2, Y_1, Y_0]):
            if (root.real > 0) and (root.imag == 0):
                y = root.real
                z = 1 + ((((u_3 * (1 - p) * y) / (
                        u_2 + (1 - p) * y)) - u_4) / r_zx)
                x = 1 + phi_xy * y ** 2 - phi_xz * z
                sols_real.append((x, y, z))
                pass
            pass
        pass
    return sols_real


def get_stable_equilibria(equilibria, parameter_values):
    r_yx, r_zx, p, phi_xy, phi_yx, phi_xz, u_1, u_2, u_3, u_4 = parameter_values
    for (i, equilibrium) in enumerate(equilibria):
        x, y, z = equilibrium
        j_11 = 1 - 2 * x + phi_xy * y ** 2 - phi_xz * z
        j_12 = 2 * phi_xy * x * y
        j_13 = -phi_xz * x
        j_21 = 2 * phi_yx * r_yx * x * y
        j_22 = r_yx * (1 - 2 * y + phi_yx * x ** 2) - (
                (u_1 * u_2 * (1 - p) * z) / (u_2 + (1 - p) * y) ** 2)
        j_23 = -((u_1 * (1 - p) * y) / (u_2 + (1 - p) * y))
        j_32 = (u_2 * u_3 * (1 - p) * z) / ((u_2 + (1 - p) * y) ** 2)
        j_33 = r_zx * (1 - 2 * z) - u_4 + (
                (u_3 * (1 - p) * y) / (u_2 + (1 - p) * y))
        C_2 = -j_11 - j_22 - j_33
        C_1 = j_11 * j_22 + j_11 * j_33 + j_22 * j_33 - j_12 * j_21 - j_23 * j_32
        C_0 = (j_11 * j_23 * j_32 + j_12 * j_21 * j_33) - (
                j_11 * j_22 * j_33 + j_13 * j_21 * j_32)
        cond1 = C_2 > 0
        cond2 = C_1 > 0
        cond3 = C_0 > 0
        cond4 = C_2 * C_1 > C_0
        if not all([cond1, cond2, cond3, cond4]):
            del equilibria[i]
            pass
        pass
    return equilibria
